## Python Scripts to Send Emails
>> python has `smtplib` library that helps to create and send an email via python scripts

In [32]:
import smtplib

**`Step 1`: create SMTP object for a server. We will use gmail server for now.**

In [45]:
# Syntax: smptlib.SMTP('server_domain_name', 'port_number')

smtp_object = smtplib.SMTP('smtp.gmail.com',587)

**`Step 2`: Run the ehlo() command which greets the server and establishes the connection.**

In [46]:
# Run the ehlo() command which greets the server and establishes the connection.
smtp_object.ehlo()

(250,
 b'smtp.gmail.com at your service, [34.125.218.159]\nSIZE 35882577\n8BITMIME\nSTARTTLS\nENHANCEDSTATUSCODES\nPIPELINING\nCHUNKING\nSMTPUTF8')

- we see first item in a tuple is `250`. It means connection is established.
- we used 587 port earlier, it means we are using TLS encryption.

**`Step 3`: Initiate TLS encryption.**

In [47]:
# Initiate TLS encryption

smtp_object.starttls()

(220, b'2.0.0 Ready to start TLS')

## Pre-requisites For Gmail Users
- You need to generate an app password instead of your normal gmail password.
- **Steps:**
  1. Enable 2 step authentication
  2. Follow steps: [click here](https://support.google.com/accounts/answer/185833?hl=en/)
  3. Copy the 16 digit app password. We will use this to send an email.

**`Step 4`: Login to your email address. Use `getpass` library to get user input.**

In [38]:
# Login to your email address. Use `getpass` library to get user input.
# Note: always use app password instead of gmail password

import getpass

email = getpass.getpass("Enter your email: ")
app_password = getpass.getpass("Enter your generated app password: ")

smtp_object.login(email, app_password)

Enter your email: ··········
Enter your generated app password: ··········


(235, b'2.7.0 Accepted')

**`Step 5:` send email using .sendmail() method**

In [40]:
# Step 5: send email using .sendmail() method

from_address = input("Enter your email (sender):")
to_address = input("Enter the email of the recipient:")

# get email subject
subject = input("Enter the subject line: ")

# get email message
message = input("Type out the message you want to send:\n")

# format message
final_message = "Subject: " + subject + '\n' + message

# send email
smtp_object.sendmail(from_address, to_address, final_message)

print("\n Message Delivered...")

Enter your email (sender):thapaliyanish123@gmail.com
Enter the email of the recipient:anish.thapaliya@fusemachines.com
Enter the subject line: demo
Type out the message you want to send:
just a demo
Message Delivered...


**`Step 6:` Close your session**

In [29]:
# close your session
smtp_object.quit()

(221,
 b'2.0.0 closing connection 6-20020a170902c10600b001a52c38350fsm1473465pli.169 - gsmtp')


<hr>
<h3>Congratulations, You can now automate sending emails
<hr>